### ISSUE TITLE
catkin build fails to find package if its only a <run_depend>

### ISSUE THREAD

**Issue by kodyensley** [catkin/catkin_tools#224]
My group is considering catkin_tools (we use 'standard' catkin now). After running the quickstart for catkin_tools, (http://catkin-tools.readthedocs.org/en/latest/quick_start.html#tl-dr), I now have four packages in my workspace, pkg_a, pkg_b, pkg_c, pkg_d. As per the guide, pkg_d has a build AND run dependency on pkg_a and pkg_b. (This all builds as expected.)

My issue comes when I modify the package.xml for pkg_d. 
```
<buildtool_depend>catkin</buildtool_depend>
<build_depend>pkg_a</build_depend>
<!--  <build_depend>pkg_b</build_depend> -->
<run_depend>pkg_a</run_depend>
<run_depend>pkg_b</run_depend>
```
This now says that pkg_a is a build and run dependency for pkg_d, but that pkg_b is only a run dependency. 

After running <b>catkin clean --all</b> for a fresh start, I run <b>catkin build pkg_d</b> and am greeted with the following error:
```
user@computer:/tmp/catkin_workspace_test$ catkin build pkg_d 
Creating build space directory, '/tmp/catkin_workspace_test/build' 
------------------------------------------------------------------------------------------------
Profile:                     default
Extending:             [env] /opt/ros/indigo
Workspace:                   /tmp/catkin_workspace_test
Source Space:       [exists] /tmp/catkin_workspace_test/src
Build Space:        [exists] /tmp/catkin_workspace_test/build
Devel Space:       [missing] /tmp/catkin_workspace_test/devel
Install Space:     [missing] /tmp/catkin_workspace_test/install
DESTDIR:                     None
------------------------------------------------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
------------------------------------------------------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
------------------------------------------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
------------------------------------------------------------------------------------------------
Workspace configuration appears valid.
------------------------------------------------------------------------------------------------ 
Found '4' packages in 0.0 seconds. 
Starting ==> pkg_a                                                                                                                                                 
[pkg_a] ==> '/tmp/catkin_workspace_test/build/pkg_a/build_env.sh /usr/bin/cmake /tmp/catkin_workspace_test/src/pkg_a -DCATKIN_DEVEL_PREFIX=/tmp/catkin_workspace_test/devel -DCMAKE_INSTALL_PREFIX=/tmp/catkin_workspace_test/install' in '/tmp/catkin_workspace_test/build/pkg_a' 
-- The C compiler identification is GNU 4.8.4                                                                                                                      
-- The CXX compiler identification is GNU 4.8.4                                                                                                                    
-- Check for working C compiler: /usr/bin/cc                                                                                                                       
-- Check for working C compiler: /usr/bin/cc -- works                                                                                                              
-- Detecting C compiler ABI info                                                                                                                                   
-- Detecting C compiler ABI info - done                                                                                                                            
-- Check for working CXX compiler: /usr/bin/c++                                                                                                                    
-- Check for working CXX compiler: /usr/bin/c++ -- works                                                                                                           
-- Detecting CXX compiler ABI info                                                                                                                                 
-- Detecting CXX compiler ABI info - done                                                                                                                          
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_workspace_test/devel                                                                                                     
-- Using CMAKE_PREFIX_PATH: /tmp/catkin_workspace_test/devel;/opt/ros/indigo                                   
-- This workspace overlays: /tmp/catkin_workspace_test/devel;/opt/ros/indigo                                                                  
-- Found PythonInterp: /usr/bin/python (found version "2.7.6")                                                                                                     
-- Using PYTHON_EXECUTABLE: /usr/bin/python                                                                                                                        
-- Using Debian Python package layout                                                                                                                              
-- Using empy: /usr/bin/empy                                                                                                                                       
-- Using CATKIN_ENABLE_TESTING: ON                                                                                                                                 
-- Call enable_testing()                                                                                                                                           
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_workspace_test/build/pkg_a/test_results                                                                              
-- Looking for include file pthread.h                                                                                                                              
-- Looking for include file pthread.h - found                                                                                                                      
-- Looking for pthread_create                                                                                                                                      
-- Looking for pthread_create - not found                                                                                                                          
-- Looking for pthread_create in pthreads                                                                                                                          
-- Looking for pthread_create in pthreads - not found                                                                                                              
-- Looking for pthread_create in pthread                                                                                                                           
-- Looking for pthread_create in pthread - found                                                                                                                   
-- Found Threads: TRUE                                                                                                                                             
-- Found gtest sources under '/usr/src/gtest': gtests will be built                                                                                                
-- Using Python nosetests: /usr/bin/nosetests-2.7                                                                                                                  
-- catkin 0.6.14                                                                                                                                                   
-- Configuring done                                                                                                                                                
-- Generating done                                                                                                                                                 
-- Build files have been written to: /tmp/catkin_workspace_test/build/pkg_a                                                                                        
[pkg_a] <== '/tmp/catkin_workspace_test/build/pkg_a/build_env.sh /usr/bin/cmake /tmp/catkin_workspace_test/src/pkg_a -DCATKIN_DEVEL_PREFIX=/tmp/catkin_workspace_test/devel -DCMAKE_INSTALL_PREFIX=/tmp/catkin_workspace_test/install' finished with return code '0' 
[pkg_a] ==> '/tmp/catkin_workspace_test/build/pkg_a/build_env.sh /usr/bin/make --jobserver-fds=3,4 -j' in '/tmp/catkin_workspace_test/build/pkg_a'                 
[pkg_a] <== '/tmp/catkin_workspace_test/build/pkg_a/build_env.sh /usr/bin/make --jobserver-fds=3,4 -j' finished with return code '0'                               
Finished <== pkg_a [ 1.4 seconds ]                                                                                                                                 
Starting ==> pkg_d                                                                                                                                                 
[pkg_d] ==> '/tmp/catkin_workspace_test/build/pkg_d/build_env.sh /usr/bin/cmake /tmp/catkin_workspace_test/src/pkg_d -DCATKIN_DEVEL_PREFIX=/tmp/catkin_workspace_test/devel -DCMAKE_INSTALL_PREFIX=/tmp/catkin_workspace_test/install' in '/tmp/catkin_workspace_test/build/pkg_d' 
-- The C compiler identification is GNU 4.8.4                                                                                                                      
-- The CXX compiler identification is GNU 4.8.4                                                                                                                    
-- Check for working C compiler: /usr/bin/cc                                                                                                                       
-- Check for working C compiler: /usr/bin/cc -- works                                                                                                              
-- Detecting C compiler ABI info                                                                                                                                   
-- Detecting C compiler ABI info - done                                                                                                                            
-- Check for working CXX compiler: /usr/bin/c++                                                                                                                    
-- Check for working CXX compiler: /usr/bin/c++ -- works                                                                                                           
-- Detecting CXX compiler ABI info                                                                                                                                 
-- Detecting CXX compiler ABI info - done                                                                                                                          
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_workspace_test/devel                                                                                                     
-- Using CMAKE_PREFIX_PATH: /tmp/catkin_workspace_test/devel;/opt/ros/indigo                                   
-- This workspace overlays: /tmp/catkin_workspace_test/devel;/opt/ros/indigo                                                                  
-- Found PythonInterp: /usr/bin/python (found version "2.7.6")                                                                                                     
-- Using PYTHON_EXECUTABLE: /usr/bin/python                                                                                                                        
-- Using Debian Python package layout                                                                                                                              
-- Using empy: /usr/bin/empy                                                                                                                                       
-- Using CATKIN_ENABLE_TESTING: ON                                                                                                                                 
-- Call enable_testing()                                                                                                                                           
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_workspace_test/build/pkg_d/test_results                                                                              
-- Looking for include file pthread.h                                                                                                                              
-- Looking for include file pthread.h - found                                                                                                                      
-- Looking for pthread_create                                                                                                                                      
-- Looking for pthread_create - not found                                                                                                                          
-- Looking for pthread_create in pthreads                                                                                                                          
-- Looking for pthread_create in pthreads - not found                                                                                                              
-- Looking for pthread_create in pthread                                                                                                                           
-- Looking for pthread_create in pthread - found                                                                                                                   
-- Found Threads: TRUE                                                                                                                                             
-- Found gtest sources under '/usr/src/gtest': gtests will be built                                                                                                
-- Using Python nosetests: /usr/bin/nosetests-2.7                                                                                                                  
-- catkin 0.6.14                                                                                                                                                   
CMake Error at /opt/ros/indigo/share/catkin/cmake/catkinConfig.cmake:75 (find_package):                                                                            
  Could not find a package configuration file provided by "pkg_b" with any of                                                                                      
  the following names:                                                                                                                                             
    pkg_bConfig.cmake                                                                                                                                              
    pkg_b-config.cmake                                                                                                                                             
  Add the installation prefix of "pkg_b" to CMAKE_PREFIX_PATH or set                                                                                               
  "pkg_b_DIR" to a directory containing one of the above files.  If "pkg_b"                                                                                        
  provides a separate development package or SDK, be sure it has been                                                                                              
  installed.                                                                                                                                                       
Call Stack (most recent call first):                                                                                                                               
  CMakeLists.txt:7 (find_package)                                                                                                                                  
-- Configuring incomplete, errors occurred!                                                                                                                        
See also "/tmp/catkin_workspace_test/build/pkg_d/CMakeFiles/CMakeOutput.log".                                                                                      
See also "/tmp/catkin_workspace_test/build/pkg_d/CMakeFiles/CMakeError.log".                                                                                       
[pkg_d] <== '/tmp/catkin_workspace_test/build/pkg_d/build_env.sh /usr/bin/cmake /tmp/catkin_workspace_test/src/pkg_d -DCATKIN_DEVEL_PREFIX=/tmp/catkin_workspace_test/devel -DCMAKE_INSTALL_PREFIX=/tmp/catkin_workspace_test/install' failed with return code '1' 
Failed   <== pkg_d [ 1.4 seconds ]                                                                                                                                 
[build] There were '1' errors:                                                                                                                                     

Failed to build package 'pkg_d' because the following command:

# Command to reproduce:
cd /tmp/catkin_workspace_test/build/pkg_d && /tmp/catkin_workspace_test/build/pkg_d/build_env.sh /usr/bin/cmake /tmp/catkin_workspace_test/src/pkg_d -DCATKIN_DEVEL_PREFIX=/tmp/catkin_workspace_test/devel -DCMAKE_INSTALL_PREFIX=/tmp/catkin_workspace_test/install; cd -

# Path to log:
cat /tmp/catkin_workspace_test/build/build_logs/pkg_d.log

Exited with return code: 1 
                                                                                                                                                                   
[build] Runtime: 2.8 seconds
```

Catkin should find pkg_b and build it, correct? This is the behavior I have come to expect from the <b>--only-pkg-with-deps</b> flag. (And when I clean the workspace and run <b>catkin_make --only-pkg-with-deps</b>, pkg_b does indeed build).

Am I missing something or is this a bug? Thanks!

**Comment by wjwwood** [comment]
This is expected behavior. The precondition for building a package is that all of its build and test dependencies have already been built. This makes sense I think. It doesn't make sense to me that the tool should build the run dependencies of a package before building it. Since those dependencies are used for running the result. If you need it at build time then it should either be a build or test dependency.

As for why this works in `catkin_make`, I believe it because `catkin_make` will run CMake over all packages and then run `make`. When you specify `--only-pkg-with-deps` it still runs CMake once over all packages and then narrows what it runs `make` on to just the package you specified and its dependencies. So this is yet another by product of having a single CMake invocation which hides a bug, in this case a missing build dependency between your packages.

Hopefully that clears things up, but feel free to ask more questions here. Please close the issue if you think it resolves your question.

**Comment by dustingooding** [comment]
I believe this is similar/same to my ticket #148.

Philosophically, I agree that pkg_d doesn't need pkg_b to *build*, but it is needed to *run*.  If you don't build a package's run depends, what good is building it (it can't run...)?  And, doesn't this break the ability to build a metapackage (all its dependencies are run/exec_depends).

The alternative, as I assume you're suggesting, is to manually call "catkin build pkg_b" before building pkg_d.  But if that's the case, why list "run_depends" at all (except maybe for Debianization later?).  And not only for pkg_b in this trivial example, but every run_depends in the entire dependency tree.  And not just before it runs, but before it builds (see OP's error).  He can't even build pkg_d until pkg_b is available, even though it's only listed as a run_depends.

I can't understand how this is a desirable or expected behavior.   A) only-run_depends shouldn't affect building, but they obviously do.  B) run_depends *should* be built (or an option should be provided to allow this, ala #148) so that built packages are useable, without needing intimate knowledge of an entire workspace's dependency tree.

**Comment by wjwwood** [comment]
> The alternative, as I assume you're suggesting, is to manually call "catkin build pkg_b" before building pkg_d. But if that's the case, why list "run_depends" at all (except maybe for Debianization later?).

I would say you should do that **after** building pkg_d. If you need to build pkg_b before you can build pkg_d then it is, by definition, a build dependency (or test dependency if it is optionally needed for testing).

The reason for not building run depends before a package is that on the farm, your run depends will not even be installed before it trys to build your package.

> A) only-run_depends shouldn't affect building, but they obviously do.

If a dependency affects the build then it is actually a build (or test) dependency. Period. I don't think there's any other way to interpret that.

> run_depends should be built (or an option should be provided to allow this, ala #148) so that built packages are useable

I think you're right. run depends to get considered in the ordering already, just not at the last level as you've described. For example, if you have `pkg_a` which build depends on `pkg_b` which run depends on `pkg_c`, then `pkg_c` is guaranteed to be built before `pkg_a`. The reason for this is that `pkg_a`'s build dependency on `pkg_b` might include using headers from `pkg_c` or running code (Python script for example) in `pkg_b` which might use stuff from `pkg_c`.

One further step would be to have an option (maybe it should be the default) which builds the given packages and their run dependencies. That would be as if you were building a package which build depends on the packages you've given. However, consider an extension of the previous example where `pkg_a` also run depends on `pkg_d`. Then I'd say the build order should be (if possible): `pkg_b`/`pkg_c`, then the other, then `pkg_a` then `pkg_d`, i.e. `pkg_a` should still be built before `pkg_d`.

**Comment by dustingooding** [comment]
I think we're in mostly violent agreement.  The ability to build run_depends seems like a desirable thing for developers (though maybe not so much for build servers).  And I agree that run_depends should be built *after* the package-in-question (as described in your last example).  But, in an attempt to not hijack this issue with my own, back to OP's issue.

I suspect this issue is not credible.  It looks like OP never removed pkg_b from pkg_d's CMakeLists.txt find_package(catkin ...) line.  So, essentially, he has an improperly defined build dependency (looked for in CMakeLists, not named in package.xml).  I was misinterpreting this issue previously, assuming that simply defining pkg_b as a run_dependency caused pkg_d to fail to build (which reminded me of my issue).  I vote for closing this issue.

**Comment by jbohren** [comment]
The fundamental problem here is why `run_depend` [has been replaced](http://www.ros.org/reps/rep-0140.html#removing-run-depend). `catkin_tools` should order based on `run_depend` deps as well, in order to not break things.

**Comment by NikolausDemmel** [comment]
@jbohren, do you mean there is something wrong with the current order of packages used by carkin_tools? As I understood it, in this issue the problem is a missing build depend.

**Comment by jbohren** [comment]
Closed with #276
